### ISSUE TITLE
Remove URI from Invoke-RestMethod's Verbose Output (Optionally)

### ISSUE THREAD

**Issue by tommymaynard** [PowerShell/PowerShell#14417]
## Summary of the new feature/enhancement

I recommend that URIs are either removed from Invoke-RestMethod's verbose output, or that they can be, according to the developer's _preference_. There is the possibility that a URI should be secure and therefore, shouldn't appear in any verbose output. If it is considered secure, this is immediately lost, if for instance it's retrieved from Azure Key Vault and  then ends up in someone's verbose output.

![image](https://user-images.githubusercontent.com/12057058/102109677-8c963900-3df1-11eb-8fc2-53f5e439a66b.png)

## Proposed technical implementation details (optional)

Removing a URI from the verbose output would likely be the easiest option. Others may prefer URIs remain in the command's verbose output, however. Therefore, perhaps there should be a mechanism/variable that displays/doesn't display the default Invoke-RestMethod's verbose output. Moving forward, maybe there should be something bigger here. It's possible that this isn't the only command that may output information that a developer doesn't want shared. Perhaps this thought/design could be future-proofed, so that other information shown in verbose output, can be removed.

**Comment by iSazonov** [comment]
Why do you turn on Verbose if you have the concern?

**Comment by tommymaynard** [comment]
I'm not the only one that invokes the functions I author. Therefore, if my code were to access Azure Key Vault (for example), obtain the URI, assign it to an environment variable, run Invoke-RestMethod using the environment variable, as the Uri parameter value, then it's going to be exposed by Write-Verbose, as I can't remove the environment variable from memory/the session until the command has completed. Therefore, it can easily be written to the PowerShell hosting program.

I do understand that the environment variable would briefly exist where the user can obtain it, if they knew what they were doing. Still, it would prove helpful in this instance to be able to control what is/isn't written by Invoke-RestMethod. _Perhaps_ I should just control whether Invoke-RestMethod can write verbose commands or not (Verbose:$false). Either way, I wanted to put this out there for discussion; thank you for your time and consideration.

**Comment by iSazonov** [comment]
Since it is Azure scenario I ping @TravisEz13 and @PaulHigin who could advice what is a best practice in the scenario and should we fix anything in the repo.

**Comment by TravisEz13** [comment]
Please be sure to follow our [Security Issue Reporting Policy](https://github.com/PowerShell/PowerShell/security/policy) in the future.

**Comment by TravisEz13** [comment]
I'm reporting and I'll lock the conversation until a determination is made.

**Comment by TravisEz13** [comment]
I have a fix for a portion of the issue we decided to fix our self after reviewing the issue.  We can resume discussion about things like optionally adding back writing the URI to the verbose stream.

Thanks for your patients.
@tommymaynard, Thanks for reporting the issue.

**Comment by tommymaynard** [comment]
Hi @TravisEz13. I appreciate the work here; however, I'm unsure what was completed that would triage my original concern. I saw the changes made **daxian-dbw**; however, I'm interested in how I control the URI/verbose output. I've installed 7.2.0-preview.3. Was there a change made in this version? I still see the URI in the verbose output, but perhaps there's more I need to know in order to experience any changes. Anything you can share with me would be great and thank you for the attention to this issue!

**Comment by TravisEz13** [comment]
This has not been released yet.

**Comment by TravisEz13** [comment]
If you want to add a new switch to add the URI back to the verbose output, feel free to submit an PR.

**Comment by jhoneill** [comment]
It seems this was implemented with very little discussion. 

The original request was to have some option which - in the case of a badly designed API which puts secure information in the URI -  doesn't show it.  If you're dealing with such URIs you put -Verbose:$false whenever they are called and the issue goes away.  
 
Instead this makes life harder for anyone who builds REST URIs. And since I do a lot of that, this change would waste hours of time. It took 5 minutes to work round by using a proxy function,  (see https://gist.github.com/jhoneill/f8ddd4e4e0a25c22d77749166d6f14fe ) , so as "security", frankly it's not doing much. 
And anyone who goes down the proxy route can headers with authorization info to verbose. 

The issue has never been closed, can we please judge this change to be a mistake and revert to the previous behaviour ?

**Comment by iSazonov** [comment]
I suggest to write limited information in verbose stream and original URL only in **trace**. 

We could parse url with URI class and build output string for verbouse as:
```powershell
$a= [uri]::new('https://anonymous:flabada@developer.mozilla.org/en-US/docs/Web/API/URL')
$a

AbsolutePath   : /en-US/docs/Web/API/URL
AbsoluteUri    : https://anonymous:flabada@developer.mozilla.org/en-US/docs/Web/API/URL
LocalPath      : /en-US/docs/Web/API/URL
Authority      : developer.mozilla.org
HostNameType   : Dns
IsDefaultPort  : True
IsFile         : False
IsLoopback     : False
PathAndQuery   : /en-US/docs/Web/API/URL
Segments       : {/, en-US/, docs/, Web/…}
IsUnc          : False
Host           : developer.mozilla.org
Port           : 443
Query          :
Fragment       :
Scheme         : https
OriginalString : https://anonymous:flabada@developer.mozilla.org/en-US/docs/Web/API/URL
DnsSafeHost    : developer.mozilla.org
IdnHost        : developer.mozilla.org
IsAbsoluteUri  : True
UserEscaped    : False
UserInfo       : anonymous:flabada

$outputString = $a.Scheme + "://" + $a.Host
$outputString

https://developer.mozilla.org

**Comment by jhoneill** [comment]
@iSazonov   That's only half way to being helpful 

Look at this way  
* If someone runs  ` iwr "http://something/secret" -verbose`   the verbose isn't the problem because the secret part is in the history anyway. 
* If a command does `$uri  = $host + $param1 + $secret_param`  then `inovke-webRequest -verbose:$false` stops someone setting `$verbosePreference`  and the secret leaking. 
* If I have code which does something like  
```
$restObj1 = Invoke-RestMethod $Uri1
$uri2 = $host + $restobj1.Links.Versions -replace "{.*$",$versionno)
```

 I want to see that the 500 error I'm getting back from the serve is because I didn't pass in the right version number.   Telling me I called the right server is zero help.  

To me this has solved a non-problem by breaking functionality which people needed.  IWR and IRM don't use debug, but that's another story. I have the proxy function as a solution but I'm currently faced with going to every place where I have ever used either of these and putting `If version >= 7.2 Write-Verbose / write-debug  $uri `

**Comment by iSazonov** [comment]
@jhoneill For regular scenarios we can show only host name. For developer scenarios I suggest to add tracing with full URL in the cmdlets.

**Comment by jhoneill** [comment]
> @jhoneill For regular scenarios we can show only host name. For developer scenarios I suggest to add tracing with full URL in the cmdlets.

Sorry I don't understand.  You *can* show show the full URL, because you did it before and I put an `Invoke-RestMethod` *function* between the rest of my code and the broken `Invoke-WebRequest` *cmdlet* to show it.    
If you mean "We *choose not to* show it because some developers might not put `-Verbose:$false` in a function when they have sensitive information in the URL"  true, but breaking things for all everyone because of lazy/poorly informed developers isn't a good way to go. It's doing nothing for security, as  (a) Anyone can hack the function and capture the URL before it goes into `Invoke-RestMethod` / `Invoke-WebRequest` (b)  Proxy function solution. 

Tracing with Set-PsDebug - if that was what you meant - doesn't show the information I need and is cumbersome to the point of being unusable.

**Comment by rkeithhill** [comment]
It seems the "fix" swung too far the other way.  I don't see how anyone should consider information in a URL sensitive or secret.  If I want to know what URLs your module is accessing, I'll just fire up Fiddler or Wireshark.  You're not going to hide that info from folks that want it.  Seems like `-Verbose:$false` would have been sufficient here.

**Comment by iSazonov** [comment]
I see a misunderstanding of scenarios. 
If the script is run in a controlled environment (like developer scenario) then this fix is ​​not useful. But in an uncontrolled environment (like CI), we do not get an _accidental_ leak - these are scenarios when the script output is available to third-party people.
A conclusion was - we don't want _accidental_ leak with verbose as verbose is often used to inform users about the progress of a script.

**Comment by jhoneill** [comment]
> In an uncontrolled environment (like CI), we do not get an _accidental_ leak - these are scenarios when the script output is available to third-party people. A conclusion was - we don't want _accidental_ leak with verbose as verbose is often used to inform users about the progress of a script.

Yes.  For the case where 
1. The URL embeds something sensitive, AND
2. The script author doesn't write -Verbose:false  AND
3. Verbose Preference is set to continue AND
4. Verbose messages are written to a log AND
5. The wrong people can access the log

Then and only then there is a chance of a leak. 
So to protect people who don't do their job properly (step 2) in a small set of use cases (1) in a small set of environments (3-5) useful functionality has been taken away from all users in all use cases / environments ?  
Or from my point of view, I'm being hindered in my job, to protect people don't do theirs properly.  If I'm being grossly unfair, feel free to correct me.

**Comment by kevinoid** [comment]
It appears that #14788 was reverted by #15853, so Uri is again included in the verbose stream in v7.3.0-preview.1 and later.  However, it appears to be inadvertent, since I don't see any mention of this issue or #14788 in #15853.

**Comment by CarloToso** [comment]
@iSazonov No longer an issue, we can close

**Comment by CarloToso** [comment]
@iSazonov the uri is no longer leaked, can we close this issue?

**Comment by iSazonov** [comment]
Fixed in #17357
